Fix #338914: Add Medieval/Renaissance clefs

Jojo-Schmitz opened the pull request:
Resolves: [#338914](https://musescore.org/en/node/338914)
Resolves: #31554

Jojo-Schmitz committed:
Fix #338914: Add Medieval/Renaissance clefs

Resolves: [#338914](https://musescore.org/en/node/338914)
Resolves: #31554

Jojo-Schmitz commented:
@rettinghaus anything more needed for MEI?

XiaoMigros commented:
Like I mentioned in the linked issue, I think allowing custom symbols for clefs is a much more sensible fix than adding an endless list of clef types.

Jojo-Schmitz commented on src/engraving/types/types.h:
```diff
@@ -596,6 +596,13 @@ enum class ClefType : signed char {
     C4_8VB,
     G8_VB_C,
+    G_PETR,
+    F_PETR,
```
oops, wrong order...

XiaoMigros commented:
Yes, which is the same for these renaissance clefs as for existing ones...
StaffType's lineoffset as well as clef Y-offset exist as well.

Jojo-Schmitz commented on src/palette/internal/palettecreator.cpp:
```diff
@@ -1075,7 +1075,8 @@ PalettePtr PaletteCreator::newClefsPalette(bool defaultPalette)
         ClefType::PERC,
-        ClefType::PERC2, ClefType::TAB, ClefType::TAB4, ClefType::TAB_SERIF, ClefType::TAB4_SERIF
+        ClefType::PERC2, ClefType::TAB, ClefType::TAB4, ClefType::TAB_SERIF, ClefType::TAB4_SERIF,
+        ClefType::G_PETR, ClefType::C1_PETR, ClefType::C2_PETR, ClefType::C3_PETR, ClefType::C4_PETR, ClefType::C5_PETR, ClefType::F_PETR
```
We may want to have them before the percussion and tablature clefs in the (more) palette

rettinghaus commented on src/importexport/mei/internal/meiconverter.cpp:
```diff
@@ -890,6 +891,11 @@ libmei::Clef Convert::clefToMEI(engraving::ClefType clef)
+    case (engraving::ClefType::C2_PETR):
+    case (engraving::ClefType::C3_PETR):
+    case (engraving::ClefType::C4_PETR):
+    case (engraving::ClefType::C5_PETR):
```
That's not a proper solution here. You should adjust the default case further down: just check if the `glyphName` starts with "mensural" and subtract it from the string.

Jojo-Schmitz replied:
And convert the then leading uppercase C, F, G into lowercase?

Jojo-Schmitz replied:
That'd also match the otherwise yet unhandled `mensuralGclef`, `mensuralFclef`, `mensuralCclef`, but still not `chantFclef` and `chantCclef` from [that range](https://www.w3.org/2019/03/smufl13/tables/medieval-and-renaissance-clefs.html)

Jojo-Schmitz commented on src/importexport/mei/internal/meiconverter.cpp:
```diff
@@ -882,6 +882,7 @@ libmei::Clef Convert::clefToMEI(engraving::ClefType clef)
     case (engraving::ClefType::G8_VA):
     case (engraving::ClefType::G15_MA):
     case (engraving::ClefType::G_1):
+    case (engraving::ClefType::G_PETR):
```
Pardon?

Jojo-Schmitz committed:
Different sorting in the palette

Jojo-Schmitz committed:
Rename '..._PETR' to just '..._R'

Jojo-Schmitz committed:
MEI import for the Petrucci/Renaissance clefs

Jojo-Schmitz commented on src/engraving/types/types.h:
```diff
@@ -596,6 +596,13 @@ enum class ClefType : signed char {
+    C3_PETR,
+    C4_PETR,
+    C5_PETR,
+    F_PETR,
```
done

Jojo-Schmitz committed:
Fix MEI export for the Medieval/Renaissance clefs

it'd also export "mensural[CFG]clef" and "chant[CF]clef" to MEI, even if those won't happen yet

cbjeukendrup commented:
Is there a consensus about whether we should merge this PR?

Personally, I also think we should eventually refactor clefs like suggested above. Perhaps even rather sooner than later, but on the other hand, not necessarily right now. 
In favour of merging this now, one could say that it only extends an existing problem slightly, but does not introduce any new one.

So what do we think?

Jojo-Schmitz commented:
Or even like this?
```diff
diff --git a/src/importexport/mnx/internal/shared/mnxtypesconv.cpp b/src/importexport/mnx/internal/shared/mnxtypesconv.cpp
index 4ef7ccdf8d..6ceda04cc4 100644
--- a/src/importexport/mnx/internal/shared/mnxtypesconv.cpp
+++ b/src/importexport/mnx/internal/shared/mnxtypesconv.cpp
@@ -89,6 +89,7 @@ std::optional<mnx::part::Clef::Required> toMnxClef(ClefType clefType)
     switch (clefType) {
     case ClefType::G:
     case ClefType::G_1:
+    case ClefType::G_R:
         sign = ClefSign::GClef;
         break;
     case ClefType::G8_VB:
@@ -115,6 +116,7 @@ std::optional<mnx::part::Clef::Required> toMnxClef(ClefType clefType)
     case ClefType::F_C:
     case ClefType::F_F18C:
     case ClefType::F_19C:
+    case ClefType::F_R:
         sign = ClefSign::FClef;
         break;
     case ClefType::F8_VB:
@@ -145,6 +147,11 @@ std::optional<mnx::part::Clef::Required> toMnxClef(ClefType clefType)
     case ClefType::C1_F20C:
     case ClefType::C3_F20C:
     case ClefType::C4_F20C:
+    case ClefType::C1_R:
+    case ClefType::C2_R:
+    case ClefType::C3_R:
+    case ClefType::C4_R:
+    case ClefType::C5_R:
         sign = ClefSign::CClef;
         break;
     case ClefType::C4_8VB:

```

Jojo-Schmitz committed:
Add Medieval/Renaissance clefs to mnx too
